File: app/src/main/java/com/security/passwordmanager/data/DataRepository.java
package com.security.passwordmanager.data;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class DataRepository {

    private final DataDao dataDao;

    public DataRepository(DataDao dataDao) {
        this.dataDao = dataDao;
    }

    public void addData(Data data) {
        if (data.isWebsite())
            dataDao.addWebsite((Website) data);
        else
            dataDao.addBankCard((BankCard) data);
    }


    public void updateData(Data data) {
        if (data.isWebsite())
            dataDao.updateWebsite((Website) data);
        else
            dataDao.updateBankCard((BankCard) data);
    }


    public List<Data> getDataList() {

        List<Website> websiteList = dataDao.getWebsiteList();
        List<BankCard> bankCardList = dataDao.getBankCardList();

        if (websiteList == null)
            websiteList = new ArrayList<>();
        if (bankCardList == null)
            bankCardList = new ArrayList<>();

        check(websiteList);
        check(bankCardList);

        return sortedConcat(websiteList, bankCardList);
    }

    @Nullable
    public List<Data> getAccountList(String key, @Data.DataType int type) {
        List<Website> websiteList = dataDao.getAccountList(key);

        if (type == Data.TYPE_BANK_CARD) {
            List<BankCard> bankCardList = dataDao.getBankAccountList(key);
            return sortedConcat(websiteList, bankCardList);
        }
        else if (websiteList == null)
            return null;
        else
            return new ArrayList<>(websiteList);
    }

    @Nullable
    public List<Data> searchData(String query) {
        if (query == null || query.length() == 0)
            return getDataList();

        List<Website> websiteList = dataDao.searchWebsite(query);
        List<BankCard> bankCardList = dataDao.searchBankCard(query);

        check(websiteList);
        check(bankCardList);

        return sortedConcat(websiteList, bankCardList);
    }


    //удаляет только 1 запись в бд
    public void deleteData(Data data) {
        if (data.isWebsite())
            dataDao.deleteWebsite((Website) data);
        else
            dataDao.deleteBankCard((BankCard) data);
    }

    //удаляет несколько записей в бд
    public void deleteData(String key, @Data.DataType int type) {
        if (type == Data.TYPE_BANK_CARD)
            dataDao.deleteBankCard(key);
        else
            dataDao.deleteWebsite(key);
    }





    private List<Data> sortedConcat(
            List<Website> websiteList, List<BankCard> bankCardList) {

        if (websiteList.size() == 0)
            return new ArrayList<>(bankCardList);
        else if (bankCardList.size() == 0)
            return new ArrayList<>(websiteList);

        List<Data> dataList = new ArrayList<>();

        int min = Math.min(websiteList.size(), bankCardList.size());
        int w = 0, b = 0;

        while (w < min && b < min) {
            Website website = websiteList.get(w);
            BankCard bankCard = bankCardList.get(b);

            if (website.compareTo(bankCard)) {
                checkAndAdd(dataList, website);
                w++;
            }
            else {
                checkAndAdd(dataList, bankCard);
                b++;
            }
        }

        if (w > b) for (; b < bankCardList.size(); b++)
            checkAndAdd(dataList, bankCardList.get(b));

        else if (w < b) for (; w < websiteList.size(); w++)
            checkAndAdd(dataList, websiteList.get(w));

        return dataList;
    }

    private void checkAndAdd(List<Data> list, Data value) {
        for (Data d : list)
            if (value.equals(d))
                return;

        list.add(value);
    }

    private void check(List<? extends Data> list) {
        int first = 0;

        while (first < list.size()) {
            Data data = list.get(first);

            for (int j = first + 1; j < list.size(); j++)
                if (data.equals(list.get(j))) {
                    list.remove(first);
                    break;
                }

            first++;
        }
    }
}

File: app/src/main/java/com/security/passwordmanager/data/DataRepository.kt
package com.security.passwordmanager.data

import com.security.passwordmanager.model.*
import com.security.passwordmanager.slice
import kotlin.math.min

class DataRepository(private val dataDao: DataDao) {

    fun addData(data: Data) = when (data) {
        is Website -> dataDao.addWebsite(data)
        is BankCard -> dataDao.addBankCard(data)
        else -> TODO()
    }


    fun updateData(data: Data) = when (data) {
        is Website -> dataDao.updateWebsite(data)
        is BankCard -> dataDao.updateBankCard(data)
        else -> TODO()
    }


    suspend fun getDataUIList(email: String, dataType: DataType? = null): List<DataUI> =
        when (dataType) {
            DataType.Website -> dataDao.getWebsiteList(email)
            DataType.BankCard -> dataDao.getBankCardList(email)
            null -> sortedConcatenation(
                dataDao.getWebsiteList(email),
                dataDao.getBankCardList(email)
            )
        }.toDataUIList()


    suspend fun searchData(email: String, query: String?, type: DataType? = null) : List<DataUI> {
        if (query == null || query.isBlank())
            return getDataUIList(email, type)

        return when (type) {
            DataType.Website -> dataDao.searchWebsite(email, query)
            DataType.BankCard -> dataDao.searchBankCard(email, query)
            null -> {
                val search = dataDao.search(email, query)
                sortedConcatenation(websiteList = search.first, bankCardList = search.second)
            }
        }.toDataUIList()
    }

    //удаляет только 1 запись в бд
    fun deleteData(data: Data) = when (data) {
        is Website -> dataDao.deleteWebsite(data)
        else -> dataDao.deleteBankCard(data as BankCard)
    }

    //удаляет несколько записей в бд
    fun deleteRecords(data: Data) = when (data) {
        is Website -> dataDao.deleteWebsite(data.email, data.key)
        else -> dataDao.deleteBankCard(data.email, data.key)
    }




    private fun sortedConcatenation(
        websiteList: MutableList<Website>,
        bankCardList: MutableList<BankCard>
    ): List<Data> {
        if (bankCardList.isEmpty())
            return websiteList
        else if (websiteList.isEmpty())
            return bankCardList

        val resultList = ArrayList<Data>()
        val minLength = min(websiteList.size, bankCardList.size)
        var w = 0
        var b = 0

        while (w < minLength && b < minLength) {
            val website = websiteList[w]
            val bankCard = bankCardList[b]
            if (website > bankCard) {
                resultList.add(website)
                w++
            }
            else {
                resultList.add(bankCard)
                b++
            }
        }

        if (w > b)
            resultList.addAll(bankCardList.slice(fromIndex = b))
        else if (w < b)
            resultList.addAll(websiteList.slice(fromIndex = w))

        return resultList
    }


    private fun List<Data>.toDataUIList() =
        groupBy { it.key }
        .map {
            DataUI(
                title = it.value[0],
                accountList = it.value.toMutableList()
            )
        }
}